Use `v` flag instead of `u` for `pattern` RegExps

mathiasbynens opened the pull request:
This makes the `pattern` attribute more powerful, enabling the use of [RegExp set notation syntax and properties of strings](https://github.com/tc39/proposal-regexp-v-flag) in its values.

[Differences with the previous `u` flag-based behavior:](https://github.com/tc39/proposal-regexp-v-flag#how-is-the-v-flag-different-from-the-u-flag)

- **\[FEATURE\]** Previously invalid patterns now become valid, e.g.

    ```
    pattern="[\p{ASCII_Hex_Digit}--[Ff]]"
    pattern="\p{RGI_Emoji}"
    pattern="[_\q{a|bc|def}]"
    ```

- **\[BREAKING CHANGE\]** Some previously valid patterns are now errors, specifically those with a character class including either an unescaped [special character](https://arai-a.github.io/ecma262-compare/snapshot.html?pr=2418#prod-ClassSetSyntaxCharacter) `(` `)` `[` `]` `{` `}` `/` `-` `\` `|` or [a double punctuator](https://arai-a.github.io/ecma262-compare/snapshot.html?pr=2418#prod-ClassSetReservedDoublePunctuator):

    ```
    pattern="[(]"
    pattern="[)]"
    pattern="[[]"
    pattern="[{]"
    pattern="[}]"
    pattern="[/]"
    pattern="[-]"
    pattern="[|]"
    pattern="[&&]"
    pattern="[!!]"
    pattern="[##]"
    pattern="[$$]"
    pattern="[%%]"
    pattern="[**]"
    pattern="[++]"
    pattern="[,,]"
    pattern="[..]"
    pattern="[::]"
    pattern="[;;]"
    pattern="[<<]"
    pattern="[==]"
    pattern="[>>]"
    pattern="[??]"
    pattern="[@@]"
    pattern="[``]"
    pattern="[~~]"
    pattern="[_^^]"
    ```

    Throwing patterns result in `inputElement.validity.valid === true` for any input value, so the only compatibility risk is that some value/pattern combinations that would previously result in `inputElement.validity.valid === false` now result in `inputElement.validity.valid === true`.

- Other previously valid patterns still behave the same. (Other than the abovementioned features, the `v` flags only differs in behavior from the `u` flag w.r.t. case-insensitive matching, but the `pattern` attribute uses case-sensitive matching.)

Note that the breaking changes apply to somewhat esoteric edge cases that can easily be avoided. In the worst case, this could cause previously invalid input to now be considered valid (since throwing patterns result in `inputElement.validity.valid === true` for any input value). IMHO making the change is worth it given the powerful new functionality it brings, and the relatively small compatibility risk. This is reminiscent of the discussion in #439 (but in a different direction).

For context, here’s a few pointers w.r.t. when we decided to implicitly enable the `u` flag for the `pattern` attribute in the first place:

- Bug/proposal: https://www.w3.org/Bugs/Public/show_bug.cgi?id=26915
- Commit: https://github.com/whatwg/html/commit/0fa713879d902d8ada81d80aa458deaf0b284b4f
- Discussion w.r.t. backwards compatibility: #439

<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

-----------

- [x] At least two implementers are interested (and none opposed):
   * [Chrome](https://github.com/whatwg/html/pull/7908#issuecomment-1425009013) on the condition of Web compatibility
   * [Firefox](https://github.com/mozilla/standards-positions/issues/745#issuecomment-1434877094)
   * [Safari/WebKit](https://github.com/WebKit/standards-positions/issues/132#issuecomment-1432995375)
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://github.com/web-platform-tests/wpt/pull/38547
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1412729
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=pattern-v
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=pattern-v
   * Deno (only for timers, structured clone, base64 utils, channel messaging, module resolution, web workers, and web storage): N/A
   * Node.js (only for timers, structured clone, base64 utils, channel messaging, and module resolution): N/A

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/7908/infrastructure.html" title="Last updated on Feb 17, 2023, 4:42 PM UTC (2a539ee)">/infrastructure.html</a>  ( <a href="https://whatpr.org/html/7908/9f8e572...2a539ee/infrastructure.html" title="Last updated on Feb 17, 2023, 4:42 PM UTC (2a539ee)">diff</a> )
<a href="https://whatpr.org/html/7908/input.html" title="Last updated on Feb 17, 2023, 4:42 PM UTC (2a539ee)">/input.html</a>  ( <a href="https://whatpr.org/html/7908/9f8e572...2a539ee/input.html" title="Last updated on Feb 17, 2023, 4:42 PM UTC (2a539ee)">diff</a> )
<a href="https://whatpr.org/html/7908/references.html" title="Last updated on Feb 17, 2023, 4:42 PM UTC (2a539ee)">/references.html</a>  ( <a href="https://whatpr.org/html/7908/9f8e572...2a539ee/references.html" title="Last updated on Feb 17, 2023, 4:42 PM UTC (2a539ee)">diff</a> )

gibson042 commented:
> Previously valid patterns still behave the same.

I don't think this is true—for example, the `v` flag [ClassSetCharacter](https://github.com/tc39/ecma262/pull/2418/files#diff-181371b08d71216599b0acccbaabd03c306da6de142ea6275c2135810999805aR34592-R34596) requires escaping for anything in [ClassSetSyntaxCharacter](https://github.com/tc39/ecma262/pull/2418/files#diff-181371b08d71216599b0acccbaabd03c306da6de142ea6275c2135810999805aR34615-R34616) (`(` `)` `[` `]` `{` `}` `/` `-` `\` `|`) while the `u` flag [ClassAtom](https://tc39.es/ecma262/multipage/text-processing.html#prod-ClassAtom) requires escaping for only `\` and `]` and sometimes `-`. Concretely, the following patterns are all valid with `u` but seem not to be with `v`:
* `[(]`
* `[)]`
* `[[]`
* `[{]`
* `[}]`
* `[/]`
* `[-]`
* `[|]`

mathiasbynens commented:
@gibson042 Thanks for the quick review! Comment updated. Any other `u` vs. `v` differences I forgot about? (We should probably list these in a new [FAQ](https://github.com/tc39/proposal-regexp-set-notation#faq) item…)

gibson042 commented:
The other obvious examples are doubled punctuators, e.g. `[..]` or `[::]`. But that's all I can think of.

mathiasbynens commented:
Updated and added links to the relevant grammar productions in the draft spec.

mathiasbynens commented:
[Chrome 112 is set to include support for the RegExp `v` flag.](https://chromestatus.com/feature/5144156542861312) (The feature is currently “staged” i.e. gated on the `--harmony` V8 flag, but will be enabled-by-default later this week.)

Given that, **I’d like to ask for initial feedback on this proposal**. I’d be happy to write WPT tests + file implementation bugs, but wanted to double-check first if this is something worth pursuing. The functionality seems useful but perhaps the potential breakage is not worth it or warrants a different approach (like a separate attribute). Please let me know your thoughts!

WDYT @domenic @annevk @zcorpan @domfarolino?

zcorpan commented:
Have you looked at httparchive for the breaking change cases, or implemented a use counter?

annevk commented:
This generally seems like a good idea, provided it's web compatible.

mathiasbynens commented:
I’ve fixed some of the existing `<input pattern>` tests in https://github.com/web-platform-tests/wpt/pull/38278 and added coverage for the current `u` flag-based behavior in https://github.com/web-platform-tests/wpt/pull/38281. Once this last one is merged, I’ll start sending follow-up draft PRs for the `v`-specific behavior described here.

mathiasbynens commented:
> Have you looked at httparchive for the breaking change cases, or implemented a use counter?

@pthier is implementing a V8-level use counter for the cases listed above occurring in `u` RegExps: https://chromium-review.googlesource.com/c/v8/v8/+/4212393 Note that this would count any patterns that cannot be switched over to `v`, regardless of whether they occur in a `pattern` attribute value or not — so it’s a superset of the truly problematic cases.

mathiasbynens commented:
Proposed WPT tests: https://github.com/web-platform-tests/wpt/pull/38325

The compatibility risk is smaller than I previously thought, since throwing patterns result in `inputElement.validity.valid === true` for any input value. So the only impact would be that some previously invalid value/pattern combinations would now be considered valid.

mfreed7 commented:
I'm also supportive of this proposal, provided it proves to be web compatible.

annevk reviewed (commented):
> If specified, the attribute's value must match the JavaScript [Pattern](https://tc39.es/ecma262/#prod-Pattern)[+UnicodeMode, +N] production.

It seems this is being modified by the "v" proposal so this probably needs updating as well?

Also, if we want to land this before it's integrated into JavaScript proper we'll need to add references to the proposal as we have done for other JavaScript proposals.

mathiasbynens commented:
> > If specified, the attribute's value must match the JavaScript [Pattern](https://tc39.es/ecma262/#prod-Pattern)[+UnicodeMode, +N] production.
> 
> It seems this is being modified by the "v" proposal so this probably needs updating as well?

Good catch. I’ve pushed a fix.

> Also, if we want to land this before it's integrated into JavaScript proper we'll need to add references to the proposal as we have done for other JavaScript proposals.

Done as well. Thanks for your review!

annevk commented on source:
```diff
@@ -3015,6 +3015,15 @@ a.setAttribute('href', 'https://example.com/'); // change the content attribute
+    proposal. The following terms are defined there, and used in this specification: <ref
+    spec=JSREGEXPVFLAG></p>
+
+    <ul class="brief">
+     <li>The <dfn data-x-href="https://arai-a.github.io/ecma262-compare/?pr=2418&amp;id=sec-isvalidregularexpressionliteral">v</dfn> flag for regular expressions</li>
+     <li>The <dfn data-x-href="https://arai-a.github.io/ecma262-compare/?pr=2418&amp;id=sec-patterns">UnicodeSetsMode</dfn> grammatical parameter</li>
+    </ul>
```
Unless we reference these from within the text, which we don't at the moment, this isn't needed.

annevk commented on source:
```diff
@@ -50307,7 +50317,7 @@ ldh-str       = &lt; as defined in <a href="https://www.rfc-editor.org/rfc/rfc10
 
    <li><p>Let <var>regexpCompletion</var> be <span>RegExpCreate</span>(<var>pattern</var>,
-   "<code data-x="">u</code>"). <ref spec=JAVASCRIPT></p></li>
+   "<code data-x="">v</code>"). <ref spec=JAVASCRIPT></p></li>
```
```suggestion
   "<code data-x="">v</code>"). <ref spec=JSREGEXPVFLAG></p></li>
```

mathiasbynens replied:
Done. Thanks!

annevk commented:
Modulo nits I think this is okay to land although it might be prudent to wait a little longer for WebKit and Mozilla positions to become official.

mathiasbynens commented:
> Modulo nits I think this is okay to land although it might be prudent to wait a little longer for WebKit and Mozilla positions to become official.

Agreed; no need to rush this. On the Chrome side we’re waiting for the UseCounter data to come in, which could take a few weeks. I’ll keep this thread updated!

mathiasbynens commented:
Based on Chrome Beta & Canary data so far, it looks like [the `HTMLPatternRegExpUnicodeSetIncompatibilitiesWithUnicodeMode` use counter](https://chromestatus.com/metrics/feature/timeline/popularity/4463) gets hit for roughly 0.015% of page loads. Given that this number [is the upper bound](https://github.com/whatwg/html/pull/7908#issuecomment-1413747011) of the number of actual incompatibility issues, I’m hopeful we can proceed with this change without breaking the Web. Thoughts?

zcorpan commented:
SGTM

annevk commented:
The graph doesn't seem to have hit a stable point yet. Is that good enough for blink-dev? I'm personally reasonable comfortable with the risks here, but I'd rather not merge this only to have to revert later and get folks to undo their implementation work.

mathiasbynens commented:
I’m waiting to send the Blink Intent to Implement & Ship until there’s data from the Stable channel. Happy to report back here as well when that happens.

I suppose I’m just not sure what the usual order of things is — are we keeping this PR open until the Blink Intent goes out, until the implementation lands, until the implementation ships? AFAICT https://github.com/web-platform-tests/wpt/pull/38547 is blocked on this PR being merged as well.

annevk commented:
For normative changes it depends on how confident we are they will be successful. Would it work if we merge when the intent gets approved?

mathiasbynens commented:
> For normative changes it depends on how confident we are they will be successful. Would it work if we merge when the intent gets approved?

Sure.

mathiasbynens commented:
FYI, I just published the Intent to Ship: https://groups.google.com/a/chromium.org/g/blink-dev/c/gIyvMw0n2qw I suspect there’ll be some discussion on the `UseCounter` results on that thread.

mathiasbynens commented:
> Would it work if we merge when the intent gets approved?

**Update:** The Intent just got approved. How do we feel about landing the WPT tests + this spec patch now?

mathiasbynens commented:
W.r.t. the `needs-compat-analysis` label: I’ve summarized the results of my investigation here: https://groups.google.com/a/chromium.org/g/blink-dev/c/gIyvMw0n2qw/m/3XaP6hFpAgAJ More details are linked for each entry. The TL;DR is that although there are many distinct sources of UseCounter hits, I haven’t found a single case that actually constituted a compat problem.

zcorpan commented:
SGTM. I'll let @annevk do the honors.

mathiasbynens commented:
Thanks everyone!

I’ve prepared the follow-up PR for if/when the TC39 proposal gets merged into the ECMAScript spec: https://github.com/tc39/ecma262/pull/2418

poebrand commented:
> [Chrome 112 is set to include support for the RegExp `v` flag.](https://chromestatus.com/feature/5144156542861312) (The feature is currently “staged” i.e. gated on the `--harmony` V8 flag, but will be enabled-by-default later this week.)
> 
> Given that, **I’d like to ask for initial feedback on this proposal**. I’d be happy to write WPT tests + file implementation bugs, but wanted to double-check first if this is something worth pursuing. The functionality seems useful but perhaps the potential breakage is not worth it or warrants a different approach (like a separate attribute). Please let me know your thoughts!
> 
> WDYT @domenic @annevk @zcorpan @domfarolino?

Hi @mathiasbynens,

As you previously noted https://github.com/tc39/ecma262/pull/2418 has not been merged yet and https://github.com/tc39/proposal-regexp-v-flag is not part of the official ECMAScript spec yet, if I'm following the conversation correctly. In which case, why are whatwg members like Chrome enabling the `v` flag by default in their official customer facing versions? 

Per https://chromestatus.com/feature/5149507107422208, Chrome 114 has `v` flag use enabled by default for `pattern`. On our own sites, we're now seeing dozens of Regexs that are no longer valid... leading to invalid values being accepted. Shouldn't such changes only go out once a feature is officially part of the ECMAScript spec? Especially with something that is a potentially breaking change depending upon specific regexes on a site?

Also, in cases where whatwg is moving forward with potentially breaking changes... shouldn't more effort be done to announce/communicate that to the wider developer community? Most existing documentation on the web includes no mention of the `v` flag (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Bad_regexp_flag).

annevk commented:
@poebrand that is unfortunate. We did follow our own process here:

1. Once a TC39 proposal hits Stage 3 it can be shipped in browsers and is eligible for incorporation into WHATWG standards.
3. https://github.com/whatwg/html/pull/9213 has been prepared for when the proposal hits Stage 4 and gets incorporated into the main ECMAScript standard. But that's essentially cleanup.

Unfortunately we only added notifying MDN of changes eight months ago and PRs older than that haven't been updated to incorporate it. I'll try to pay better attention to that going forward, although for potentially breaking changes that might still not be sufficient I suppose. Not entirely sure what would be though.

zcorpan commented:
Filed:
- https://github.com/mdn/content/issues/27346
- https://github.com/mdn/content/issues/27347

As far as breakage, I believe the reasoning here was that websites should also do server-side validation (since users can remove the `pattern` attribute in devtools or submit invalid data in other ways), so the [few](https://github.com/whatwg/html/pull/7908#issuecomment-1466906703) affected sites should not break if they have server-side validation.

Kovalchik8 commented:
> Filed:
> 
> * [RegExp v flag not documented mdn/content#27346](https://github.com/mdn/content/issues/27346)
> * [<input pattern> changed to set the RegExp 'v' flag mdn/content#27347](https://github.com/mdn/content/issues/27347)
> 
> As far as breakage, I believe the reasoning here was that websites should also do server-side validation (since users can remove the `pattern` attribute in devtools or submit invalid data in other ways), so the [few](https://github.com/whatwg/html/pull/7908#issuecomment-1466906703) affected sites should not break if they have server-side validation.

My projects are using a common regexps within HTML pattern attribute found on popular forums and now almost all of them are broken in Chrome, so I doubt this incompatibility counter is reliable. Okay, I do a server-side validation, but a browser side validation is important as well since it has it's own logic and UI. Such a frustrating update. How to migrate painlessly from old regexps to new ones using /v flag?
